development/tools/pkg/gcscleaner/cleaner.go

```diff
@@ -1,118 +1,253 @@
 			bucketName),
 		)
 		return false
 	}
 	timestamp, err := strconv.ParseInt(*timestampSuffix, 32, 0)
 	if err != nil {
-		logrus.Debug(fmt.Sprintf(
+		logrus.Info(fmt.Sprintf(
 			"skipping bucket '%s', unable to parse timestamp",
 			bucketName),
 		)
 		return false
 	}
-	if now-timestamp < int64(cfg.BucketLifespanDuration) {
-		logrus.Debug(fmt.Sprintf(
+	duration := now - timestamp
+	if duration < int64(r.cfg.BucketLifespanDuration) {
+		logrus.Info(fmt.Sprintf(
 			"bucket: '%s' is %s old and will not be deleted, the duration: '%s' was not exceeded",
 			bucketName,
-			time.Duration(now-timestamp),
-			cfg.BucketLifespanDuration),
+			time.Duration(duration),
+			r.cfg.BucketLifespanDuration),
 		)
 		return false
 	}
-	logrus.Debug(fmt.Sprintf(`bucket: '%s' will be deleted`, bucketName))
+	logrus.Debug(fmt.Sprintf(`bucket: '%s' was created %s ago and will be deleted`, bucketName, time.Duration(duration)))
 	return true
 }
 
-func extractTimestampSuffix(name string, regTimestampSuffix regexp.Regexp) *string {
-	submatch := regTimestampSuffix.FindSubmatch([]byte(name))
+func (r Cleaner) extractTimestampSuffix(name string) *string {
+	submatch := r.cfg.BucketNameRegexp.FindSubmatch([]byte(name))
 	if len(submatch) < 2 {
 		return nil
 	}
 	result := string(submatch[1])
 	return &result
 }
 
-// NewConfig creates new cleaner configuration
-func NewConfig(
-	bucketNameRegexp regexp.Regexp,
-	excludedBucketNames []string,
-	bucketLifespanDuration time.Duration) Config {
+func (r Cleaner) deleteBucketObject(
+	ctx context.Context,
+	bucketName string,
+	objectName string) error {
+	msg := fmt.Sprintf("object deleted: %s", objectName)
+	if r.cfg.IsDryRun {
+		logrus.Debug(msg)
+		return nil
+	}
+	err := r.client.Bucket(bucketName).Object(objectName).Delete(ctx)
+	logrus.Debug(msg)
+	return err
+}
+
+func (r Cleaner) iterateBucketObjectNames(
+	ctx context.Context,
+	bucketName string,
+	bucketObjectChan chan storage.BucketObject,
+	errChan chan error) {
+	defer close(bucketObjectChan)
+
+	bucket := r.client.Bucket(bucketName)
+	objectIterator := bucket.Objects(ctx, nil)
+	for {
+		select {
+		case <-ctx.Done():
+			errChan <- ctx.Err()
+			return
+		default:
+			attrs, err := objectIterator.Next()
+			if err == iterator.Done {
+				return
+			}
+			if err != nil {
+				errChan <- err
+				return
+			}
+			bucketObjectChan <- storage.NewBucketObject(attrs.Bucket(), attrs.Name())
+		}
+	}
+}
 
-	return Config{
-		RegTimestampSuffix:     bucketNameRegexp,
-		ExcludedBucketNames:    excludedBucketNames,
-		BucketLifespanDuration: bucketLifespanDuration,
+func (r Cleaner) deleteBucketObjects(
+	ctx CancelableContext,
+	bucketObjectChan chan storage.BucketObject,
+	errChan chan error) {
+	for {
+		select {
+		case bo, ok := <-bucketObjectChan:
+			if !ok {
+				return
+			}
+			if err := r.deleteBucketObject(ctx, bo.Bucket(), bo.Name()); err != nil {
+				errChan <- err
+				ctx.Cancel()
+				return
+			}
+		default:
```

Comment:
Is the default needed? In such case loop will be not blocked on `case bo, ok := <-bucketObjectChan:`